### DISCUSSION TITLE
Support for consuming "Forwarded" HTTP header (RFC7239)

### DISCUSSION THREAD

**Discussion by dionysius** [dani-garcia/vaultwarden#3097] [Ideas]
Link to RFC: https://datatracker.ietf.org/doc/html/rfc7239

The `Forwarded` HTTP header quasi supersedes the technically non-standardized (de-facto standardized) `X-Forwarded-*` headers.
Using the Forwarded header allows to chain proxies and with it chaining those forwarded fields.

Suggestion:
If in the config the exact string `Forwarded` is set for `IP_HEADER`, vaultwarden will try to read the first (or last?) `for` property as the real ip.

If there really is a chain of forwarded fields, one might want to configure from which group the `for` field should be read from. If the new `IP_FORWARDED_BY` config option is set only that groups `for` property is consumed where `by` matches this fields' value.

**Comment by BlackDex** [comment]
I'm not seeing this RFC being implemented in all the main reverse proxies a.t.m.
Nginx, Apache, Caddy, Traefik, HaProxy all do not support this out-of-the-box currently.

Once that is a bit more mainstream I would gladly add support for this. But for now, I think it is best to just use the `IP_HEADER` configuration we have currently. Most reverse proxy have build-in functionality for the `X-Forwarded-For` with multiple IP's by appending them, which is save as far as i know.

**Comment by csutcliff** [comment]
HAProxy 2.8 supports this (May 2023)

**Reply by dionysius** [reply]
~~Pretty sure HAProxy supports it longer since they [kinda invented it](https://github.com/haproxy/haproxy/blob/master/doc/proxy-protocol.txt).~~ https://github.com/haproxy/haproxy/commit/b2bb9257d255e5d5377d9d3ce5b2d41795207669 Jan 2023

**Comment by dionysius** [comment]
List: https://c960657.github.io/forwarded.html
